### PR TITLE
busybox df does not support -x or -l

### PR DESCRIPTION
**Pull request by JensKuehnel**
The df part of the OpenWRT agent does nor work with busybox.
busybox does not support -x or -l. It works when both are removed.

### CODE REVIEW

**Comment by rawiriblundell** [comment]
Hi again @JensKuehnel,
As far as I can tell, the openwrt agent was originally forked from the linux agent script.  In the search to reduce maintenance overhead, there is (or was) a goal to merge all the *nix agent scripts together, so the openwrt and linux scripts are obvious first contenders to be merged.  You can reference my early proposal for this [here](https://github.com/tribe29/checkmk/pull/28/files#diff-88a095b049ca26da05ee82e4b5c8a7a02cf103223c43f083f0a0b324c0d3b7caR999).

Now that I have an OpenWRT device at my disposal, I'm in a better position to do real-world busybox testing.

The tricky part for smooshing all this disparate `df` code together is handling the filtering out of undesirable filesystem types.  Typically within a busybox environment this shouldn't be an issue, but OTOH if portable code is the goal, now's possibly a good time to test it.  As you can see in my previously proposed code, I'd gone with `grep -Ev`, IIRC just for the sake of getting that block of code finished.

One of the problems with that approach is that it matches the entire line, not just the Type field.  So I think one possible solution here is to switch to `awk`.

And, as a side note: Annoyingly, `-T` means different things on different systems: on some it means print the output with each filesystem's type, on others it means print only for the defined filesystem types.  So let's ignore handling for that case for now.

```
_df_excludefs="Type|autofs|basejail|cifs|devfs|devtmpfs|fdescfs|iso9660|mvfs"
_df_excludefs="${_df_excludefs}|nfsv4|nfs|prl_fs|procfs|smbfs|squashfs|udf"

df -kPT | awk -v filter="${_df_excludefs}" '$2 !~ filter'
```

That works for me in both OpenWRT and Linux.  This approach also conveniently absorbs `sed`'s job at the same time.

Next, my version of OpenWRT (18.06) comes with busybox 1.29.3, and isn't compiled with the `-i` flag for `df`, despite that being added to busybox's `df` source code back in 2007.  At the same time, it supports `-T`, which was added to busybox's `df` in 2014.  Go figure.

So as a tweak to your PR, I'd wrap the block that handles inodes output with a capability test like this:

```
# df inodes information
if df -i >/dev/null 2>&1; then
    printf -- '%s\n' "<<<df>>>" "[df_inodes_start]"
    df -PTi | awk -v filter="${_df_excludefs}" '$2 !~ filter'
    printf -- '%s\n'  '[df_inodes_end]'
fi
```

You may need to adjust the field that `awk` acts on.

Please give these suggestions a test and see if it works for you :)

Cheers

Rawiri

File: agents/check_mk_agent.openwrt [hunk 1]
```diff
@@ -265,15 +265,12 @@ section_df() {
     # These should generally be monitored on the NFS server, not on the clients.
 
     echo '<<<df>>>'
-    # The exclusion list is getting a bit of a problem. -l should hide any remote FS but seems
-    # to be all but working.
-    excludefs="-x smbfs -x cifs -x iso9660 -x udf -x nfsv4 -x nfs -x mvfs -x zfs -x prl_fs"
-    df -PTlk $excludefs | sed 1d
+    df -PTk | sed 1d
```
Comment (jherbel) [inline agents/check_mk_agent.openwrt:268]: Please incorporate the filtering suggested by @rawiriblundell

File: agents/check_mk_agent.openwrt [hunk 2]
```diff
@@ -265,15 +265,12 @@ section_df() {
...
 
     # df inodes information
     echo '<<<df>>>'
     echo '[df_inodes_start]'
-    df -PTli $excludefs | sed 1d
+    df -PTi | sed 1d
```
Comment (jherbel) [inline agents/check_mk_agent.openwrt:273]: Please incorporate the filtering suggested by @rawiriblundell

**Review by jherbel — CHANGES_REQUESTED** [review]
Thanks for the PR. This should be fixed, however, simply removing the filtering is not a good solution.

**Comment by jherbel** [comment]
Hi @rawiriblundell , thanks for your suggestions. To make sure that I understand correctly: `sed 1d` used to remove the first line, right? Your suggestion includes this removal by filtering (among others) against "Type" with `awk`, is that correct? One thing is not clear to me: 
```
_df_excludefs="Type|autofs|basejail|cifs|devfs|devtmpfs|fdescfs|iso9660|mvfs"
_df_excludefs="${_df_excludefs}|nfsv4|nfs|prl_fs|procfs|smbfs|squashfs|udf"
```
This exclude list contains more than we previously attempted to exclude with `-x` option (e.g. "basejail" or "squashfs"). Could you please briefly elaborate on this? I.e., did you simply extend the list with things that seemed reasonable?

**Comment by rawiriblundell** [comment]
> Hi @rawiriblundell , thanks for your suggestions. To make sure that I understand correctly: `sed 1d` used to remove the first line, right? Your suggestion includes this removal by filtering (among others) against "Type" with `awk`, is that correct?

Hi @jherbel, that's correct.  `awk` is matching (or not-matching, in this case) on the second field, which we can see here is the word "Type":

```
▓▒░$ df -kPT
Filesystem     Type     1024-blocks      Used Available Capacity Mounted on
```

`-P` is POSIX/portability format, so these are the fields that we can reasonably expect to match on in the header of `df`.  In other words: this should be a robust test that can safely replace `sed 1d`.

>One thing is not clear to me:
> 
> ```
> _df_excludefs="Type|autofs|basejail|cifs|devfs|devtmpfs|fdescfs|iso9660|mvfs"
> _df_excludefs="${_df_excludefs}|nfsv4|nfs|prl_fs|procfs|smbfs|squashfs|udf"
> ```
> 
> This exclude list contains more than we previously attempted to exclude with `-x` option (e.g. "basejail" or "squashfs"). Could you please briefly elaborate on this? I.e., did you simply extend the list with things that seemed reasonable?

Sure, no problem :)  In my previous comment, I referenced my old PR 28 merged script code [here](https://github.com/tribe29/checkmk/pull/28/files#diff-88a095b049ca26da05ee82e4b5c8a7a02cf103223c43f083f0a0b324c0d3b7caR999).  You can see the approach that I took there was to take the different `df` sections from each existing script, put them together in an all-encompassing function and select the approach based on an environment variable: `$MK_OSSTR`.

(Any attempt to merge the scripts together will require that variable, by the way.  If I choose to try committing code again, that would be one of the first PR's I'd put in.)

In this instance, however, I took the opportunity to merge in exclusions from the respective AIX and FreeBSD code-blocks.  In other words, this exclude list covers Linux and OpenWRT... and proactively lays some ground-work for more easily merging in AIX and FreeBSD.  The chances of a filesystem type named "basejail" appearing on an OpenWRT router?  Laughable :)  So in my view, it's a safe and forward thinking move.  (BTW: squashfs is from the Linux script ;) )

In fact, it looks like with just a tiny little bit more effort, we'd be able to fully sync up `section_df()` across the openwrt and linux scripts, right here, right now.

Hope that clears that up?

Cheers

Rawiri

**Comment by jherbel** [comment]
Ok, thanks, this makes sense. It is true that this PR lays some groundwork for unifying things. However, we should definitely stay within the scope of the PR, so @JensKuehnel , please stick so the exact exclusion list we had before (this also makes it easier to understand this change later on). Unifying things should be done in separate commits later on.

**Comment by JensKuehnel** [comment]
Hi again ;-),
OpenWRT version 18.06 is unsupported since December 2020. An update to 19.07 or 21.02 is needed to use a version with security updates available. You can install OpenWRT also in a Virtual Machine on x86_64. I do some of my testing there.
My ARM Hardware (access Point) is running 21.02, but I did some test on 19.07 in my x86_64 VM. (Hint use the ext4 version, you can not update it (like on ARM), but the squashfs version is so small you can not install anything.

I changed the scripts as discussed, but for the Linux Client it should use -x, because otherwise you could get a timeout when nfs is not available. The awk solution is not a good idea than.

CU
Jens

**Comment by rawiriblundell** [comment]
> Hi again ;-), OpenWRT version 18.06 is unsupported since December 2020. An update to 19.07 or 21.02 is needed to use a version with security updates available. You can install OpenWRT also in a Virtual Machine on x86_64. I do some of my testing there.

Hi @JensKuehnel ,
I was previously testing in a most-recent-version VM (and using busybox on Linux for reference), but this is a good example of one of the gotchas for scripting for the checkmk agents:  You often find yourself having to make concessions and code against your actual wishes in order support the oddities present in older versions, because the reality is that old OS's and old software is still out there in production around the world.  It can be a bit of a burden, but at the same time it can produce some interesting challenges.

>but for the Linux Client it should use -x, because otherwise you could get a timeout when nfs is not available

Let's leave that out-of-scope for now :)

@jherbel: looks good to me, FWIW.  Over to you...

Cheers

Rawiri

**Comment by jherbel** [comment]
Hi, there is currently some internal discussion going on whether this is the right approach. The alternative would be to not do any filtering at all here and instead leave that to a parse function, which would probably be the cleaner approach. We were also wondering if on OpenWRT, `df` is only available via BusyBox or if it also available from other sources which might support `-x` and `-l`.

**Comment by JensKuehnel** [comment]
Hi,
there is a coreutils-df package available, but this is not the default. If someone creates an custom image, it is no problem to include it in the build. For people who use the precreated images, you could install it as a package, but during an upgrade this packages are removed. So you should not request coreutils-df to be installed, to use check_mk.

**Comment by jherbel** [comment]
Ok, thanks for the reply. Could you please upload an agent output with all the df filtering removed, i.e. with this in the agent:
```
echo '<<<df>>>'
df -kPT

# df inodes information
if df -i >/dev/null 2>&1; then
    echo '<<<df>>>'
    echo '[df_inodes_start]'
    df -PTi
    echo '[df_inodes_end]'
fi
```

**Comment by jherbel** [comment]
Hi @JensKuehnel , are you still interested in this?

**Comment by JensKuehnel** [comment]
Hi @jherbel ,
sorry to much other things to do. I pushed the change as requested and tested it on my setup. Works as well.

**Comment by jherbel** [comment]
Hi @JensKuehnel , thanks for the changes. As I said, could you please upload/post a corresponding agent output? The df-sections (everything under <<<df>>>, include any inode information) would be enough. Thanks!

**Comment by JensKuehnel** [comment]
Hi,
@jherbel  here the output of the agent from openwrt 21.02.1 on TP-Link Archer C7 v2.

CU
Jens

<<<df>>>
Filesystem         Type     1024-blocks  Used Available Capacity Mounted on
/dev/root          squashfs        5632  5632         0     100% /rom
tmpfs              tmpfs          61700   240     61460       1% /tmp
/dev/mtdblock4     jffs2           8640   460      8180       6% /overlay
overlayfs:/overlay overlay         8640   460      8180       6% /
tmpfs              tmpfs            512     0       512       0% /dev
<<<df>>>
[df_inodes_start]
Filesystem         Type     Inodes IUsed IFree IUse% Mounted on
/dev/root          squashfs   1383  1383     0  100% /rom
tmpfs              tmpfs     15425    97 15328    1% /tmp
/dev/mtdblock4     jffs2         0     0     0     - /overlay
overlayfs:/overlay overlay       0     0     0     - /
tmpfs              tmpfs     15425    59 15366    1% /dev
[df_inodes_end]

**Comment by jherbel** [comment]
Thank you very much. The changes were merged and this PR will be closed shortly. Please note that we decided to add a timeout to the `df` calls to avoid hangs in case of remote filesystems (since we cannot filter with `-x` or `-l`). This means that the `df` section will only be available if the `waitmax` command is available (either via the `timeout` command or via the `waitmax` binary provided by Checkmk).

**Pull request closed without merging**